src/Surfnet/StepupGateway/SecondFactorOnlyBundle/Controller/SecondFactorOnlyController.php

```diff
@@ -0,0 +1,261 @@
+            ->setRelayState($httpRequest->get(AuthnRequest::PARAMETER_RELAY_STATE, ''))
+            ->setResponseAction('SurfnetStepupGatewaySecondFactorOnlyBundle:SecondFactorOnly:respond')
+            ->setResponseContextServiceId(static::RESPONSE_CONTEXT_SERVICE_ID);
+
+        // Check if the NameID is provided and we may use it.
+        $nameId = $originalRequest->getNameId();
+        if (!$this->verifyNameId($originalRequest->getServiceProvider(), $nameId, $logger)) {
+            /** @var \Surfnet\StepupGateway\SecondFactorOnlyBundle\Service\ResponseRenderingService $responseRendering */
+            $responseRendering = $this->get('second_factor_only.response_rendering');
+            return $responseRendering->renderRequesterFailureResponse(
+                $this->getResponseContext()
+            );
+        }
+        $stateHandler->saveIdentityNameId($nameId);
+
+        // Check if the requested Loa is provided and supported.
+        $authnContextClassRef = $originalRequest->getAuthenticationContextClassRef();
+        $loaId = $this->verifyAuthnContextClassRef($authnContextClassRef, $logger);
+        if (!$loaId) {
+            /** @var \Surfnet\StepupGateway\SecondFactorOnlyBundle\Service\ResponseRenderingService $responseRendering */
+            $responseRendering = $this->get('second_factor_only.response_rendering');
+            return $responseRendering->renderRequesterFailureResponse(
+                $this->getResponseContext()
+            );
+        }
+        $stateHandler->setRequiredLoaIdentifier($loaId);
+
+        $logger->notice(
+            'Forwarding to second factor controller for loa determination and handling'
+        );
+        return $this->forward(
+            'SurfnetStepupGatewayGatewayBundle:SecondFactor:selectSecondFactorForVerification'
+        );
+    }
+
+    /**
+     * @param string $authnContextClassRef
+     * @param LoggerInterface $logger
+     * @return string
+     */
+    private function verifyAuthnContextClassRef(
+        $authnContextClassRef,
+        LoggerInterface $logger
+    ) {
+        if (!$authnContextClassRef) {
+            $logger->info(
+                'No LOA requested, sending response with status Requester Error'
+            );
+            return;
+        }
+
+        /** @var LoaAliasLookupService $loaAliasLookup */
+        $loaAliasLookup = $this->get('second_factor_only.loa_alias_lookup');
+        $loaId = $loaAliasLookup->findLoaIdByAlias($authnContextClassRef);
+
+        if (!$loaId) {
+            $logger->info(sprintf(
+                'Requested required Loa "%s" does not have a second factor alias,'
+                .' sending response with status Requester Error',
+                $authnContextClassRef
+            ));
+            return;
+        }
+
+        $loaResolutionService = $this->get('surfnet_stepup.service.loa_resolution');
+
+        if (!$loaResolutionService->hasLoa($loaId)) {
```

Comment:
I think this whole method could be refactored into a service, which uses the LoA alias look-up and LoA resolution services.